arbnode/transaction_streamer.go

```diff
@@ -1292,113 +1291,96 @@ func (s *TransactionStreamer) executeMessages(ctx context.Context, ignored struc
+	s.espressoTxnsStateInsertionMutex.Lock()
+	defer s.espressoTxnsStateInsertionMutex.Unlock()
+
 	submittedTxns, err := s.getEspressoSubmittedTxns()
 	if err != nil {
-		return fmt.Errorf("submitted pos not found: %w", err)
+		return fmt.Errorf("submitted transactions not found: %w", err)
 	}
 	if len(submittedTxns) == 0 {
 		return nil // no submitted transaction, treated as successful
 	}
 
-	firstSubmitted := submittedTxns[0]
-	hash := firstSubmitted.Hash
-
-	submittedTxHash, err := tagged_base64.Parse(hash)
-	if err != nil || submittedTxHash == nil {
-		return fmt.Errorf("invalid hotshot tx hash, failed to parse hash %s: %w", hash, err)
-	}
+	newSubmittedTxns := []arbutil.SubmittedEspressoTx{}
+	for _, submittedTx := range submittedTxns {
+		hash := submittedTx.Hash
+		submittedTxHash, err := tagged_base64.Parse(hash)
+		if err != nil || submittedTxHash == nil {
+			return fmt.Errorf("invalid hotshot tx hash, failed to parse hash %s: %w", hash, err)
+		}
 
-	data, err := s.espressoClient.FetchTransactionByHash(ctx, submittedTxHash)
-	if err != nil {
-		return fmt.Errorf("unable to fetch transaction by hash: %w", err)
-	}
-	height := data.BlockHeight
+		data, err := s.checkEspressoQueryNodesForTransasction(ctx, submittedTxHash)
+		if err != nil {
+			resubmittedTxn, err := s.resubmitTransactionIfPastDelay(ctx, submittedTx)
+			if err != nil {
+				log.Error("failed to resubmit transaction", "err", err)
+				continue
+			}
+			newSubmittedTxns = append(newSubmittedTxns, *resubmittedTxn)
+			continue
+		}
 
-	jsonHeader, err := s.espressoClient.FetchRawHeaderByHeight(ctx, height)
-	if err != nil {
-		return fmt.Errorf("could not get the header (height: %d): %w", height, err)
-	}
+		height := data.BlockHeight
 
-	var header espressoTypes.HeaderImpl
-	err = json.Unmarshal(jsonHeader, &header)
-	if err != nil {
-		return fmt.Errorf("could not unmarshal header from bytes (height: %d): %w", height, err)
-	}
+		resp, err := s.espressoClient.FetchTransactionsInBlock(ctx, height, s.chainConfig.ChainID.Uint64())
+		if err != nil {
+			log.Warn("Failed to fetch transactions in block referenced in fetch transaction by hash", "height", height, "error", err)
+			continue
+		}
 
-	log.Info("Fetching Merkle Root at hotshot", "height", height)
-	// Verify the merkle proof
-	snapshot, err := s.lightClientReader.FetchMerkleRoot(height, nil)
-	if err != nil {
-		return fmt.Errorf("%w (height: %d): %w", EspressoValidationErr, height, err)
-	}
+		validated := arbutil.ValidateIfPayloadIsInBlock(submittedTx.Payload, resp.Transactions)
+		if !validated {
+			log.Warn("Transaction payload not found in block, attempt to resubmit", "height", height, "tx", submittedTx.Hash)
+			resubmittedTxn, err := s.resubmitTransactionIfPastDelay(ctx, submittedTx)
+			if err != nil {
+				log.Error("failed to resubmit transaction", "err", err)
+				continue
+			}
+			newSubmittedTxns = append(newSubmittedTxns, *resubmittedTxn)
```

Comment:
So now that we have included the `resubmittedTxn` should we remove the original transaction?

Reply:
So now the logic is like:

- Create a new empty submitted transactions array
- Poll finality for every submitted transaction
- If finality is not found:
    - if it has reached to the deadline: resubmit and push this resumitted txn to this new array
    - if not, push this submitted to the new array
- If finiality is seen, then the submitted txn should not be pushed to the new array
- Persist the new submitted txns array

Reply:
okay makes sense!